### PR TITLE
Dfr 4984 in 0149184   email address invalid for intervener and data validation error

### PR DESCRIPTION
**Pull request by hamzahtahirhmcts**
### Jira link
[DFR-4984](https://tools.hmcts.net/jira/browse/DFR-4984)

### Change description

**Enhanced Intervener Data Validation:**

- Added stricter validation for intervener solicitor email addresses (must be present, valid, and registered to an organization for MyHMCTS access).
- Improved postcode validation for interveners. The postcode is now a required field, and clearer error messages are provided.
- Refactored validation logic (moved/checks centralized into ContactDetailsValidator).
- Updated IntervenersAboutToSubmitHandler to include and enforce the new validation checks before intervener data is updated/saved.

**Service and Handler Updates:**

- Injected ValidatePartiesService in handlers/services for use in new email/org validation logic.
- Improved logging and error messaging, including more specific errors if email addresses are not registered.
- Updated how solicitor roles are revoked if representation changes and the previous email/org were invalid.

**Test Coverage Increased and Updated:**

- Added/updated unit tests in IntervenersAboutToSubmitHandlerTest to cover all new validation scenarios, including:
- Valid and invalid solicitor emails.
- Missing/blank postcodes.
- Role revocation correctness.
- Relevant parametrized/edge case coverage added.
- Added new test cases into ContactDetailsValidatorTest for intervener solicitor email and organization validation.
- Updated IntervenerServiceTest to assert on new validation rules and error messaging.

**Other Improvements:**

- Various method signatures updated to support enhanced validation and error handling.
- Consistent formatting/comment improvements for maintainability and clarity.

### Security Vulnerability Assessment ###

**CVE Suppression:** Are there any CVEs present in the codebase (either newly introduced or pre-existing) that are being intentionally suppressed or ignored by this commit?
  * [ ] Yes
  * [ ] No

### Checklist

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerServiceTest.java [hunk 1]
```diff
@@ -1795,7 +1810,9 @@ void givenCase_whenRemoveIntervenerWithoutStoredUserIdAndUserNotFound_thenLogErr
         verify(systemUserService).getSysUserToken();
         verify(organisationService).findUserByEmail(INTERVENER_TEST_EMAIL, AUTH_TOKEN);
         verify(assignCaseAccessService, never()).removeCaseRoleToUser(any(), any(), any(), any());
-        assertThat(errors).contains("Could not find intervener with provided email");
+        assertThat(errors).contains(
+            INTERVENER_TEST_EMAIL + " is not a valid Email address. The email address must be registered to access MyHMCTS"
+        );
```
Comment (hamzahtahirhmcts) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerServiceTest.java:1791]: NOTE FOR DEVS

This test file ought to be updated. This file is almost hitting 2000 lines! Worth refactoring, I will create a ticket for this on the tech debt backlog.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/intervener/IntervenerService.java [hunk 2]
```diff
@@ -156,19 +161,52 @@ private void validateIntervenerCountryOfResident(IntervenerWrapper intervenerWra
...
+        //If there is no previous intervener/intervener solicitor
+        if (ObjectUtils.isEmpty(beforeIntv) || !isRepresented(beforeIntv)) {
+            return;
+        }
+
+        String previousIntervenerSolEmail = beforeIntv.getIntervenerSolEmail();
+
+        //Only revoke access to previous emails if it's valid with org. No need to stop case from processing.
+        String error = ContactDetailsValidator.checkForIntervenerSolicitorEmailAddress(beforeIntv, validatePartiesService);
+        if (!StringUtils.hasText(error)) {
```
Comment (hamzahtahirhmcts) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/intervener/IntervenerService.java:208]: FYI @al-hmcts 

This should prevent the production case from attempting to revoke the previous invalid email address of the intervener. We only revoke valid emails that are connected to an org.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/helper/ContactDetailsValidator.java
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/helper/ContactDetailsValidator.java:1]: remove unused variable as well `INVALID_VALIDATE_POSTCODE_METHOD_MESSAGE`

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: Return value of the method is never used

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java [hunk 3]
```diff
@@ -156,19 +161,52 @@ private void validateIntervenerCountryOfResident(IntervenerWrapper intervenerWra
...
-                                                          String email, List<String> errors) {
+                                                          String email,
+                                                          List<String> errors) {
         FinremCaseData beforeData = caseDetailsBefore.getData();
         IntervenerWrapper beforeIntv = intervenerWrapper.getIntervenerWrapperFromCaseData(beforeData);
 
-        if (ObjectUtils.isNotEmpty(beforeIntv)
-            && isRepresented(beforeIntv)) {
+        //If there is no previous intervener/intervener solicitor
+        if (ObjectUtils.isEmpty(beforeIntv) || !isRepresented(beforeIntv)) {
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: To move `ObjectUtils.isEmpty(beforeIntv)` logic into `isRepresented`

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/intervener/IntervenerService.java [hunk 4]
```diff
@@ -156,19 +161,52 @@ private void validateIntervenerCountryOfResident(IntervenerWrapper intervenerWra
...
             String beforeOrgId = beforeIntv.getIntervenerOrganisation().getOrganisation().getOrganisationID();
-            if (ObjectUtils.notEqual(beforeOrgId, orgId) || !beforeIntv.getIntervenerSolEmail().equals(email)) {
-                revokeIntervenerRole(caseDetailsBefore.getId(), beforeIntv.getIntervenerSolEmail(),
+
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/intervener/IntervenerService.java:210]: It's better to have null checking on `beforeIntv.getIntervenerOrganisation().getOrganisation().getOrganisationID();`. Use `Optional.ofNullable` and `map` if possible.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java [hunk 5]
```diff
@@ -156,19 +161,52 @@ private void validateIntervenerCountryOfResident(IntervenerWrapper intervenerWra
...
+        String previousIntervenerSolEmail = beforeIntv.getIntervenerSolEmail();
+
+        //Only revoke access to previous emails if it's valid with org. No need to stop case from processing.
+        String error = ContactDetailsValidator.checkForIntervenerSolicitorEmailAddress(beforeIntv, validatePartiesService);
+        if (!StringUtils.hasText(error)) {
             String beforeOrgId = beforeIntv.getIntervenerOrganisation().getOrganisation().getOrganisationID();
-            if (ObjectUtils.notEqual(beforeOrgId, orgId) || !beforeIntv.getIntervenerSolEmail().equals(email)) {
-                revokeIntervenerRole(caseDetailsBefore.getId(), beforeIntv.getIntervenerSolEmail(),
+
+            if (ObjectUtils.notEqual(beforeOrgId, orgId) || !previousIntervenerSolEmail.equals(email)) {
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: Use `OrganisationPolicy.isSameOrganisation` if possible. but you'll need to change the parameter of `checkIfIntervenerSolicitorDetailsChanged`

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/intervener/IntervenerService.java [hunk 6]
```diff
@@ -156,19 +161,52 @@ private void validateIntervenerCountryOfResident(IntervenerWrapper intervenerWra
...
+            if (ObjectUtils.notEqual(beforeOrgId, orgId) || !previousIntervenerSolEmail.equals(email)) {
+                revokeIntervenerRole(
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/intervener/IntervenerService.java:218]: could you resolve the deprecated codes in  `revokeIntervenerRole`?

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/intervener/IntervenersAboutToSubmitHandlerTest.java [hunk 7]
```diff
@@ -48,288 +50,222 @@ class IntervenersAboutToSubmitHandlerTest {
...
-    private IntervenersAboutToStartHandler aboutToStartHandler;
-
-    @InjectMocks
-    private IntervenersMidHandler midHandler;
-
     @Mock
     private IntervenerService intervenerService;
 
+    @Mock
+    private ValidatePartiesService validatePartiesService;
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/intervener/IntervenersAboutToSubmitHandlerTest.java]: it's not used. I believe some conditions may not be covered.

Comment (hamzahtahirhmcts) [reply src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/intervener/IntervenersAboutToSubmitHandlerTest.java]: [47d2e56](https://github.com/hmcts/finrem-case-orchestration-service/pull/2929/commits/47d2e560dda35d302015eae18c6b3328e2c4c22c)

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/intervener/IntervenerService.java (hunk 6 above)
Comment (hamzahtahirhmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/intervener/IntervenerService.java:218]: [b04b848](https://github.com/hmcts/finrem-case-orchestration-service/pull/2929/commits/b04b8484601bfa4e6ff914113b699fb45996896e)

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java (hunk 5 above)
Comment (hamzahtahirhmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: [e2f0abc](https://github.com/hmcts/finrem-case-orchestration-service/pull/2929/commits/e2f0abc52e044bf2da104390bc5e2e8122cfbfd8)

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/intervener/IntervenerService.java (hunk 4 above)
Comment (hamzahtahirhmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/intervener/IntervenerService.java:210]: [565798a](https://github.com/hmcts/finrem-case-orchestration-service/pull/2929/commits/565798a9cc81eed3faad134f22f5741e8964e6f7)

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java (hunk 3 above)
Comment (hamzahtahirhmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: [6993c77](https://github.com/hmcts/finrem-case-orchestration-service/pull/2929/commits/6993c77a88bed48698f6816cf59135dbc3547158)

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/helper/ContactDetailsValidator.java [hunk 8]
```diff
@@ -275,6 +275,36 @@ public static List<String> validateOrganisationPolicy(FinremCaseData caseData) {
...
+     * @param intervener the intervener whose solicitor email address is to be validated
+     * @param validatePartiesService the service used for validation to check whether the email is registered
+     * @return a validation error message if the email address is missing or invalid; otherwise {@code null}
+     */
+    public static String checkForIntervenerSolicitorEmailAddress(IntervenerWrapper intervener,
+                                                                 ValidatePartiesService validatePartiesService) {
+
+        String intervenerSolicitorEmail = intervener.getIntervenerSolEmail();
+        if (YesOrNo.YES.equals(intervener.getIntervenerRepresented())) {
+            if (!isValidEmailAddress(intervener.getIntervenerSolEmail())) {
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/helper/ContactDetailsValidator.java]: use `intervenerSolicitorEmail`

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: remove systemUserService

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java [hunk 9]
```diff
@@ -92,7 +98,7 @@ public IntervenerChangeDetails removeIntervenerDetails(IntervenerWrapper interve
         return intervenerChangeDetails;
     }
 
-    public IntervenerChangeDetails updateIntervenerDetails(IntervenerWrapper intervenerWrapper,
+    public void updateIntervenerDetails(IntervenerWrapper intervenerWrapper,
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: update indentation

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java [hunk 10]
```diff
@@ -110,13 +116,13 @@ public IntervenerChangeDetails updateIntervenerDetails(IntervenerWrapper interve
         FinremCaseDetails caseDetailsBefore = callbackRequest.getCaseDetailsBefore();
         if (isRepresented(intervenerWrapper)) {
             log.info("Add {} case role for Case ID: {}", caseRole, caseId);
-            String orgId = intervenerWrapper.getIntervenerOrganisation().getOrganisation().getOrganisationID();
+            Organisation newOrganisation = intervenerWrapper.getIntervenerOrganisation().getOrganisation();
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: check getIntervenerOrganisation null ? could make use of Optional.ofNullable

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java [hunk 11]
```diff
@@ -156,22 +162,65 @@ private void validateIntervenerCountryOfResident(IntervenerWrapper intervenerWra
...
+     *
+     * <p>
+     * This method compares the current and previous intervener details. If the previous intervener was represented and had
+     * a valid solicitor email address, and either the organisation or email address has changed, the previous solicitor's
+     * case role is revoked. If the previous solicitor email is invalid, no action is taken.
+     * </p>
+     *
+     * @param intervenerWrapper   the current intervener details
+     * @param caseDetailsBefore   the previous case details for comparison
+     * @param newOrganisation               the current organisation
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: incorrect spacings

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java [hunk 12]
```diff
@@ -156,22 +162,65 @@ private void validateIntervenerCountryOfResident(IntervenerWrapper intervenerWra
...
+        if (!isRepresented(beforeIntv)) {
+            return;
+        }
+
+        String previousIntervenerSolEmail = beforeIntv.getIntervenerSolEmail();
+
+        //Only revoke access to previous emails if it's valid with org. No need to stop case from processing.
+        String error = ContactDetailsValidator.checkForIntervenerSolicitorEmailAddress(beforeIntv, validatePartiesService);
+        if (!StringUtils.hasText(error)) {
+            String beforeOrgId = getOrganisationId(beforeIntv);
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: put this after 

` if (!isSameOrganisation(newOrganisation, previousOrganisation) || !previousIntervenerSolEmail.equals(email)) {`

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java [hunk 13]
```diff
@@ -156,22 +162,65 @@ private void validateIntervenerCountryOfResident(IntervenerWrapper intervenerWra
...
+            Organisation previousOrganisation = beforeIntv.getIntervenerOrganisation().getOrganisation();
+
+            if (!isSameOrganisation(newOrganisation, previousOrganisation) || !previousIntervenerSolEmail.equals(email)) {
+                revokeIntervenerRole(
+                    caseDetailsBefore.getId(),
+                    previousIntervenerSolEmail,
                     beforeOrgId,
-                    intervenerWrapper.getIntervenerSolicitorCaseRole().getCcdCode(), errors, beforeIntv);
+                    intervenerWrapper.getIntervenerSolicitorCaseRole().getCcdCode(),
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: Use Optional.ofNullable

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java [hunk 14]
```diff
@@ -156,22 +162,65 @@ private void validateIntervenerCountryOfResident(IntervenerWrapper intervenerWra
...
+                    errors,
+                    beforeIntv
+                );
             }
+        } else {
+            log.info("Invalid previous intervener email address. No revoke needed.");
         }
     }
 
+    private String getOrganisationId(IntervenerWrapper intervenerWrapper) {
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: can this be put under `IntervenerWrapper` class? Making it as @JsonIgnore

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/intervener/IntervenersAboutToSubmitHandlerTest.java [hunk 15]
```diff
@@ -63,273 +58,210 @@ void testCanHandle() {
...
 
-        midHandler.handle(finremCallbackRequest, AUTH_TOKEN);
-
-        DynamicRadioListElement operation = DynamicRadioListElement.builder().code(ADD_INTERVENER_ONE_CODE).build();
-        finremCaseData.getIntervenerOptionList().setValue(operation);
-
-        aboutToSubmitHandler.handle(finremCallbackRequest, AUTH_TOKEN);
-        verify(intervenerService).updateIntervenerDetails(oneWrapper, new ArrayList<>(), finremCallbackRequest);
+        verify(intervenerService).updateIntervenerDetails(intervener, response.getErrors(), callbackRequest);
+        verify(intervenerService, never()).removeIntervenerDetails(any(), any(), any(), any());
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/intervener/IntervenersAboutToSubmitHandlerTest.java]: You method name told me it's just to ensure `updateIntervenerDetails` is called. So we don't need to verify `removeIntervenerDetails` is never called.

and before handle, could you enhance the code readability?
e.g. 

```
        FinremCallbackRequest callbackRequest = getValidFinremCallbackRequest(ADD_INTERVENER);

        GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> response =
            aboutToSubmitHandler.handle(callbackRequest, AUTH_TOKEN);

        assertAll(
            () -> verify(intervenerService).updateIntervenerDetails(intervener, response.getErrors(), callbackRequest),
            () -> assertThat(response.getErrors()).isEmpty()
        );
```

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/intervener/IntervenersAboutToSubmitHandler.java [hunk 16]
```diff
@@ -75,13 +80,35 @@ public GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle(Finrem
...
 
-        return GenericAboutToStartOrSubmitCallbackResponse.<FinremCaseData>builder()
-            .data(caseData).errors(errors).build();
+        return response(caseData, null, errors);
+    }
+
+    private void validateIntervenerInformation(IntervenerWrapper intervener, List<String> errors) {
+        addErrorIfPresent(
+            errors,
+            ContactDetailsValidator.checkForIntervenerSolicitorEmailAddress(intervener, validatePartiesService)
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/intervener/IntervenersAboutToSubmitHandler.java]: a unit test of this is missing. You may use MockStatic to verify validatePartiesService is passed.

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerServiceTest.java [hunk 17]
```diff
@@ -486,6 +479,8 @@ void givenContestedCase_whenUpdatingIntervener1AndChangedRepresentationSolEmailC
         OrganisationPolicy organisationPolicy = OrganisationPolicy.builder().organisation(
             Organisation.builder().organisationID(SOME_ORG_ID).organisationName(SOME_ORG_ID).build()
         ).build();
+        when(validatePartiesService.isEmailRegisteredInOrg(INTERVENER_TEST_EMAIL, SOME_ORG_ID))
```
Comment (so99y) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerServiceTest.java]: put it under @BeforeEach method so you don't need to repeat it.

```
@BeforeEach
void setup() {
    lenient().when(validatePartiesService.isEmailRegisteredInOrg(INTERVENER_TEST_EMAIL, SOME_ORG_ID));
}
```

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/intervener/IntervenersAboutToSubmitHandler.java [hunk 18]
```diff
@@ -75,13 +80,35 @@ public GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle(Finrem
             throw new IllegalArgumentException("Invalid operation code: " + selectedOperationCode);
         }
 
-        return GenericAboutToStartOrSubmitCallbackResponse.<FinremCaseData>builder()
-            .data(caseData).errors(errors).build();
+        return response(caseData, null, errors);
+    }
+
+    private void validateIntervenerInformation(IntervenerWrapper intervener, List<String> errors) {
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/intervener/IntervenersAboutToSubmitHandler.java]: I suggest `validateIntervenerInformation` could be moved to `IntervenerService`.

Comment (so99y) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/intervener/IntervenersAboutToSubmitHandler.java]: Therefore, in the handler unit test, the validation could simply be mocked instead of building a sophisticated callback request class.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java (hunk 10 above)
Comment (hamzahtahirhmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/IntervenerService.java]: It should not be null as the organisation is required from the user for the new organisation. I don't think it's necessary unless there's a possibility?

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/intervener/IntervenerService.java [hunk 19]
```diff
@@ -227,6 +275,34 @@ public IntervenerChangeDetails setIntervenerRemovedChangeDetails(IntervenerWrapp
         return intervenerChangeDetails;
     }
 
+    public void validateIntervenerInformation(IntervenerWrapper intervener, List<String> errors) {
```
Comment (so99y) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/intervener/IntervenerService.java:298]: JavaDoc

**Review by so99y — DISMISSED** [review]

**Review by al-hmcts — DISMISSED** [review]

**Review by al-hmcts — APPROVED** [review]

**Review by so99y — APPROVED** [review]
